ros/src/simulation/gazebo_simulator/vehicle/vehicle_gazebo_simulation_interface/src/vehicle_gazebo_input_subscriber.cpp

```diff
@@ -0,0 +1,171 @@
+    void sterringAngleCallback(const std_msgs::Float64::ConstPtr &input_steering_angle_msg);
+    void velocityCallback(const std_msgs::Float64::ConstPtr &input_velocity_msg);
+    void controlCommandStampedCallback(const autoware_msgs::ControlCommandStamped::ConstPtr &input_msg);
+
+    double wheel_base_;
+    double wheel_tread_;
+    double wheel_radius_;
+
+  public:
+    VehicleGazeboInputSubscriber();
+    ~VehicleGazeboInputSubscriber(){};
+};
+
+VehicleGazeboInputSubscriber::VehicleGazeboInputSubscriber() : nh_(""), pnh_("~")
+{
+
+    pnh_.param("wheel_base", wheel_base_, 2.95);
+    pnh_.param("wheel_radius", wheel_radius_, 0.341);
+    pnh_.param("wheel_tread", wheel_tread_, 1.55);
+    bool twist_sub, twiststamped_sub, steering_angle_sub, velocity_sub, cmd_vel_sub, ctrl_cmd_sub;
+    pnh_.param("twist_sub", twist_sub, true);
+    pnh_.param("twiststamped_sub", twiststamped_sub, false);
+    pnh_.param("steering_angle_sub", steering_angle_sub, false);
+    pnh_.param("velocity_sub", velocity_sub, false);
+    pnh_.param("ctrl_cmd_sub", ctrl_cmd_sub, true);
+    wheel_right_rear_pub_ = nh_.advertise<std_msgs::Float64>("wheel_right_rear_velocity_controller/command", 1, true);
+    wheel_left_rear_pub_ = nh_.advertise<std_msgs::Float64>("wheel_left_rear_velocity_controller/command", 1, true);
+    steering_right_front_pub_ = nh_.advertise<std_msgs::Float64>("steering_right_front_position_controller/command", 1, true);
+    steering_left_front_pub_ = nh_.advertise<std_msgs::Float64>("steering_left_front_position_controller/command", 1, true);
+
+    if (twist_sub)
+        twist_sub_ = nh_.subscribe("/cmd_vel", 1, &VehicleGazeboInputSubscriber::twistCallback, this);
+    if (twiststamped_sub)
+        twiststamped_sub_ = nh_.subscribe("/twist_cmd", 1, &VehicleGazeboInputSubscriber::twistStampedCallback, this);
+    if (steering_angle_sub)
+        steering_angle_sub_ = nh_.subscribe("/steering_angle", 1, &VehicleGazeboInputSubscriber::sterringAngleCallback, this);
+    if (velocity_sub)
+        velocity_sub_ = nh_.subscribe("/velocity", 1, &VehicleGazeboInputSubscriber::velocityCallback, this);
+    if (ctrl_cmd_sub)
+        ctrl_cmd_sub_ = nh_.subscribe("/ctrl_cmd", 1, &VehicleGazeboInputSubscriber::controlCommandStampedCallback, this);
+}
+
+void VehicleGazeboInputSubscriber::twistStampedCallback(const geometry_msgs::TwistStamped::ConstPtr &input_twist_msg)
+{
+    std_msgs::Float64 output_wheel_rear, output_steering_right_front, output_steering_left_front;
+    output_wheel_rear.data = input_twist_msg->twist.linear.x / wheel_radius_;
+
+    double vref_rear = input_twist_msg->twist.linear.x;
+    if (std::fabs(vref_rear) < 0.01)
+    {
+        vref_rear = 0.0 < vref_rear ? 0.01 : -0.01;
+    }
+
+    double delta_ref = std::atan(input_twist_msg->twist.angular.z * wheel_base_ / vref_rear);
+    delta_ref = 0.0 < vref_rear ? delta_ref : -delta_ref;
+    if (M_PI / 4.0 < std::fabs(delta_ref))
```

Comment:
What is `4.0`?

Reply:
It is a constraint that the theory does not turn more than 90 degrees

Reply:
eb9cf22a8f4a4ca76341822b0ec7f910f28b49ea

Reply:
d77454e8fe4ca3f5bea3233cf5c11cd185db0bca